lib/Common/Common/CompressionUtilities.cpp

```diff
@@ -0,0 +1,138 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "CommonCommonPch.h"
+#include "CompressionUtilities.h"
+
+#ifdef ENABLE_COMPRESSION_UTILITIES
+#include <compressapi.h>
+#endif
+
+#define IFFALSEGO(expr,label) do { if(!(expr)) { goto label; } } while(0);
+
+using namespace Js;
+
+DWORD ConvertCompressionAlgorithm(CompressionUtilities::CompressionAlgorithm algorithm)
+{
+    // Note: The algorithms listed in CompressionUtilities.h should be kept in-sync with those
+    // defined in compressapi.h or else we will need to do more than a simple cast here.
+    return static_cast<DWORD>(algorithm);
+}
+
+HRESULT CompressionUtilities::CompressBuffer(
+    _In_ ArenaAllocator* alloc,
+    _In_ const byte* inputBuffer,
+    _In_ size_t inputBufferByteCount,
+    _Out_ byte** compressedBuffer,
+    _Out_ size_t* compressedBufferByteCount,
+    _In_opt_ CompressionAlgorithm algorithm)
+{
+    Assert(compressedBuffer != nullptr);
+    Assert(compressedBufferByteCount != nullptr);
+
+    *compressedBuffer = nullptr;
+    *compressedBufferByteCount = 0;
+
+    HRESULT hr = E_FAIL;
+
+#ifdef ENABLE_COMPRESSION_UTILITIES
+    COMPRESSOR_HANDLE compressor = nullptr;
+    if (!CreateCompressor(ConvertCompressionAlgorithm(algorithm), nullptr, &compressor))
+    {
+        return E_FAIL;
+    }
+
+    if (algorithm == CompressionAlgorithm_Xpress || algorithm == CompressionAlgorithm_Xpress_Huff)
+    {
+        DWORD level = 0;
+        IFFALSEGO(SetCompressorInformation(compressor, COMPRESS_INFORMATION_CLASS_LEVEL, &level, sizeof(DWORD)), Error);
+    }
+
+    SIZE_T compressedByteCount = 0;
+    bool result = Compress(compressor, inputBuffer, inputBufferByteCount, nullptr, 0, &compressedByteCount);
+
+    if (!result)
+    {
+        DWORD errorCode = GetLastError();
+        if (errorCode != ERROR_INSUFFICIENT_BUFFER)
+        {
+            hr = HRESULT_FROM_WIN32(errorCode);
+            goto Error;
+        }
+    }
+
+    *compressedBuffer = AnewNoThrowArray(alloc, byte, compressedByteCount);
+    IFFALSEGO(*compressedBuffer != nullptr, Error);
+
+    SIZE_T compressedDataSize;
+    IFFALSEGO(Compress(compressor, inputBuffer, inputBufferByteCount, *compressedBuffer, compressedByteCount, &compressedDataSize), Error);
+    *compressedBufferByteCount = compressedDataSize;
+
+    hr = S_OK;
+
+Error:
+    if (compressor != nullptr)
+    {
+        CloseCompressor(compressor);
```

Comment:
Are we likely to serialize a lot of parser state at once at some point (e.g. on page close, or after initial parse), in such a way that we could improve performance by not creating a new compressor for each compression? I'm not familiar with the performance characteristics of the compression library, but with the way that the API is designed it seems to imply that there's some creation and tear-down cost.

Reply:
Good question. We serialize the parser state after parsing but before executing each script so we'll create a Compressor for each script, then. The creation and tear-down cost appear minimal so I didn't look into sharing the instances but it would probably work. I guess we could lazily create Compressors/Decompressors for each algorithm and store them in a static array in CompressionUtilities. Would clean them up in ThreadContext::~ThreadContext maybe. I'm a little reluctant to do this now as the cost here doesn't seem worth the effort but may come back to this.

Reply:
👍 Yeah, that makes sense. Seems odd that they'd design the api that way if the cost is negligible though.